### PR TITLE
log-backup: added intervally resolve regions

### PR DESCRIPTION
**Pull request by YuJuncen**
Signed-off-by: hillium <yujuncen@pingcap.com>

### What is changed and how it works?

Issue Number: Ref https://github.com/tikv/tikv/issues/13638

What's Changed:

This PR is the "resolved TS" part of https://github.com/tikv/tikv/pull/14023.

```commit-message
This PR added a “two phase” flush to log backup for reducing checkpoint lag.
Generally, we added a `MinTs` task, where resolve the regions and advance the `resolved_ts` in the checkpoint manager.
then, once we are doing flush, we would make current `resolved_ts` become `checkpoint_ts`.
This allows us to advance checkpoint_ts even the leader has gone. When the leader changes frequently, this can greatly reduce checkpoint lag.
```

### Related changes

- PR to update `pingcap/docs`/`pingcap/docs-cn`:
- Need to cherry-pick to the release branch

### Check List 

Tests 

- Unit test
- Integration test

Side effects

- Performance regression
    - Consumes more CPU
    - Consumes more MEM

### Release note 

```release-note
Make the checkpoint lag of PITR more stable when there are some leadership transforming.
```

### CODE REVIEW

File: components/backup-stream/src/endpoint.rs [hunk 1]
```diff
@@ -953,6 +972,34 @@ where
...
+                                Task::RegionCheckpointsOp(RegionCheckpointOperation::Resolved {
+                                    checkpoints: resolved.take_resolve_result(),
+                                    start_time,
+                                });
+                            try_send!(sched, t);
+                        }),
+                        min_ts
+                    })
+                );
+            }
```
Comment (Leavrth) [inline components/backup-stream/src/endpoint.rs:1002]: It seems that `RegionCheckpointOperation::Resolve` is only scheduled by `RegionCheckpointOperation::PrepareMinTsForResolve` ?
```suggestion
            RegionCheckpointOperation::PrepareMinTsForResolve => {
                let min_ts = self.pool.block_on(self.prepare_min_ts());
                let start_time = Instant::now();
                let sched = self.scheduler.clone();
                try_send!(
                    self.scheduler,
                    Task::ModifyObserve(ObserveOp::ResolveRegions {
                        callback: Box::new(move |mut resolved| {
                            let t =
                                Task::RegionCheckpointsOp(RegionCheckpointOperation::Resolved {
                                    checkpoints: resolved.take_resolve_result(),
                                    start_time,
                                });
                            try_send!(sched, t);
                        }),
                        min_ts
                    })
                );
            }
```

**Review by Leavrth — APPROVED** [review]

File: components/backup-stream/src/endpoint.rs (hunk 1 above)
Comment (YuJuncen) [reply components/backup-stream/src/endpoint.rs:1002]: We must reschedule the task to consume all the pending events after we updated the concurrency manager.

Comment (Leavrth) [reply components/backup-stream/src/endpoint.rs:1002]: got it

File: components/backup-stream/src/endpoint.rs [hunk 2]
```diff
@@ -919,13 +921,30 @@ where
         }
     }
 
+    fn min_ts_worker(&self) -> future![()] {
+        let sched = self.scheduler.clone();
+        async move {
+            loop {
+                tokio::time::sleep(Duration::from_secs(10)).await;
```
Comment (3pointer) [inline components/backup-stream/src/endpoint.rs:928]: looks like it may cause too many check leader and increase cross az traffic for every 10 seconds.

Comment (3pointer) [reply components/backup-stream/src/endpoint.rs:928]: make it a configuration is better

File: components/backup-stream/src/endpoint.rs [hunk 3]
```diff
@@ -997,17 +1047,34 @@ pub enum RegionSet {
...
 impl fmt::Debug for RegionCheckpointOperation {
     fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
         match self {
-            Self::Update(arg0) => f.debug_tuple("Update").field(arg0).finish(),
+            Self::Flush => f.debug_tuple("Flush").finish(),
             Self::Get(arg0, _) => f.debug_tuple("Get").field(arg0).finish(),
+
             Self::Subscribe(_) => f.debug_tuple("Subscription").finish(),
+            Self::Resolved { checkpoints, .. } => {
+                f.debug_tuple("Resolve").field(checkpoints).finish()
```
Comment (3pointer) [inline components/backup-stream/src/endpoint.rs:1072]:
```suggestion
                f.debug_tuple("Resolved").field(checkpoints).finish()
```
BTW maybe use a clear name like "Resolving" "Resolved".

File: src/config/mod.rs [hunk 4]
```diff
@@ -2595,6 +2597,13 @@ impl BackupStreamConfig {
             );
             self.num_threads = default_cfg.num_threads;
         }
+        if self.max_flush_interval < ReadableDuration::secs(1) {
```
Comment (3pointer) [inline src/config/mod.rs:2600]: maybe also check `min_ts_interval`

**Review by 3pointer — APPROVED** [review]

**Review by hicqu — APPROVED** [review]

**Comment by YuJuncen** [comment]
/test
`[2023-03-13T04:23:05.868Z] thread 'test::resolved_follower' panicked at 'called `Option::unwrap()` on a `None` value', components/backup-stream/tests/mod.rs:545:84` 🤔

**Comment by YuJuncen** [comment]
(Perhaps we need to add a callback for `flush`...)

**Pull request merged**